lib/sidekick/sidekick.ex

```diff
@@ -1,75 +1,88 @@
+      |> Base.decode32!(padding: false)
+      |> :erlang.binary_to_term()
+
+    {:ok, _} = Application.ensure_all_started(:elixir)
+    {:ok, _} = Application.ensure_all_started(:parent)
+
+    parent_node = node(caller)
+    true = Node.connect(parent_node)
+
+    {:ok, _} = Sidekick.Supervisor.start(parent_node, children)
+
+    send(caller, {__MODULE__, :initialized})
   end
 
-  defp node_host_name(name) do
-    hostname = Node.self() |> Atom.to_string() |> String.split("@") |> List.last()
-    :"#{name}@#{hostname}"
+  defp hostname do
+    [_name, hostname] = String.split("#{node()}", "@", parts: 2)
+    hostname
   end
 
-  defp wait_for_sidekick(sidekick_node, parent_node, children) do
-    :net_kernel.monitor_nodes(true)
-    command = start_node_command(sidekick_node, parent_node)
-    Port.open({:spawn, command}, [:stream])
+  defp start_node(sidekick_node, children) do
+    command = start_node_command(sidekick_node, children)
+    port = Port.open({:spawn, command}, [:stream, :exit_status])
 
+    # Note that we're not using a timeout, because sidekick is programmed to connect to this node
+    # or self-terminate if that fails. Therefore, the situation where sidekick is running but not
+    # connected isn't possible.
     receive do
-      {:nodeup, ^sidekick_node} ->
-        # wait for node to really be up
-        # TODO deal with this in a better way
-        Process.sleep(500)
-
-        case call(sidekick_node, Sidekick.Supervisor, :start_link, [parent_node, children]) do
-          {:ok, pid} ->
-            {:ok, sidekick_node, pid}
-
-          {:error, error} ->
-            Node.spawn(sidekick_node, :init, :stop, [])
-            {:error, error}
-        end
-    after
-      5000 ->
-        # Shutdown node if we never received a response
-        Node.spawn(sidekick_node, :init, :stop, [])
-        {:error, :timeout}
+      # {:nodeup, ^sidekick_node} -> :ok
+      {__MODULE__, :initialized} -> :ok
+      {^port, {:exit_status, _status}} -> {:error, :boot_error}
     end
   end
 
-  defp start_node_command(sidekick_node, parent_node) do
+  defp start_node_command(sidekick_node, children) do
     {:ok, command} = :init.get_argument(:progname)
-    paths = Enum.join(:code.get_path(), " , ")
 
-    base_args = "-noinput -name #{sidekick_node}"
+    name_arg =
+      case :net_kernel.longnames() do
+        true -> "-name #{sidekick_node}"
+        false -> "-sname #{sidekick_node}"
+        _ -> raise "not in distributed mode"
+      end
+
+    base_args = "-noinput #{name_arg}"
 
     priv_dir = :code.priv_dir(:ci)
     boot_file_args = "-boot #{priv_dir}/node"
 
     cookie = Node.get_cookie()
     cookie_arg = "-setcookie #{cookie}"
 
-    paths_arg = "-pa #{paths}"
+    paths_args = :code.get_path() |> Enum.map(&"-pa #{&1}") |> Enum.join(" ")
 
-    command_args = "-s Elixir.Sidekick start_sidekick #{parent_node}"
+    encoded_arg = {self(), children} |> :erlang.term_to_binary() |> Base.encode32(padding: false)
```

Comment:
Ahh sweet, I had such a hard time figuring out how to do this. That I moved the init logic back into start node. This is much metter